recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,398 @@
+                    "prob_perplex": predictions["prob_perplex"],
+                    "code_perplex": predictions["code_perplex"],
+                    "num_vars": predictions["num_vars"],
+                    "temp": predictions["temp"],
+                }
+            )
+        return objectives
+
+    def fit_batch(self, batch):
+        should_step = self.step % self.grad_accumulation_factor == 0
+        # Managing automatic mixed precision
+        if self.auto_mix_prec:
+            with self.modules.model.no_sync() if not should_step else ExitStack():
+                with torch.cuda.amp.autocast():
+                    outputs = self.compute_forward(batch, Stage.TRAIN)
+                    objectives = self.compute_objectives(
+                        outputs, batch, Stage.TRAIN
+                    )
+                loss = objectives["loss"]
+                if self.hparams.diversity_loss_weight == 0.0:
+                    backprop_loss = loss
+                else:
+                    backprop_loss = (
+                        loss
+                        + objectives["diversity_loss"] * self.hparams.diversity_loss_weight# * objectives["num_masked"]
+                        + objectives["latent_l2"] * self.hparams.latent_l2_penalty_weight
+                    )
+                self.scaler.scale(
+                    backprop_loss / self.grad_accumulation_factor
+                ).backward()
+
+                objectives["total_loss"] = backprop_loss.detach()
+                if should_step:
+                    self.scaler.unscale_(self.optimizer)
+                    if self.check_gradients(loss):
+                        self.scaler.step(self.optimizer)
+                    self.scaler.update()
+                    self.optimizer_step += 1
+                    self.optimizer.zero_grad()
+        else:
+
+            with self.modules.model.no_sync() if not should_step else ExitStack():
+                outputs = self.compute_forward(batch, Stage.TRAIN)
+                objectives = self.compute_objectives(
+                    outputs, batch, Stage.TRAIN
+                )
+                loss = objectives["loss"]
+                if self.hparams.diversity_loss_weight == 0.0:
+                    backprop_loss = loss
+                else:
+                    backprop_loss = (
+                        loss
+                        + objectives["diversity_loss"] * self.hparams.diversity_loss_weight
+                        + objectives["latent_l2"] * self.hparams.latent_l2_penalty_weight
+                    )
+                (backprop_loss / self.grad_accumulation_factor).backward()
+                objectives["total_loss"] = backprop_loss.detach()
+                if should_step:
+                    self.optimizer.step()
+                    self.optimizer.zero_grad()
+                    self.optimizer_step += 1
+
+        if should_step:
+            if isinstance(
```

Comment:
Why is this needed ?

Reply:
when wrapping a module with `DistributedDataParallel` one does not have access to the module's attributes anymore, so depending on the whether DDP is used or not the API changes from `normal_module.attribute` to `ddp_module.module.attribute`

It's ugly I agree but don't know of a better approach https://discuss.pytorch.org/t/access-to-attributes-of-model-wrapped-in-ddp/130572